lib/spack/spack/filesystem_view.py

```diff
@@ -274,29 +265,96 @@ def add_standalone(self, spec):
                                         long=False)
                 return False
 
-        tree = LinkTree(spec.prefix)
-
-        if not self.ignore_conflicts:
-            conflict = tree.find_conflict(self.root)
-            if conflict is not None:
-                tty.error(self._croot +
-                          "Cannot link package %s, file already exists: %s"
-                          % (spec.name, conflict))
-                return False
+        self.merge(spec)
 
-        conflicts = tree.merge(self.root, link=self.link,
-                               ignore=ignore_metadata_dir,
-                               ignore_conflicts=self.ignore_conflicts)
         self.link_meta_folder(spec)
 
-        if self.ignore_conflicts:
-            for c in conflicts:
-                tty.warn(self._croot + "Could not link: %s" % c)
-
         if self.verbose:
             tty.info(self._croot + 'Linked package: %s' % colorize_spec(spec))
         return True
 
+    def merge(self, spec, ignore=None):
+        pkg = spec.package
+        if hasattr(pkg, 'view_source'):
```

Comment:
Rather than using `hasattr` here, add functions that return sensible defaults to `PackageBase` so that they can be documented there.  Then you don't have to check whether the attributes are present.